File: src/AS_24_01_week2/wangon/1477.cpp
// 휴게소 세우기
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <set>
#include <string>
#include <vector>
using namespace std;
int n, m, l;
vector<int> road;
int ret = 0;

bool func(int mid)
{
	int cnt = 0;
	for (int i = 1; i < road.size(); i++)
	{
		int distance = road[i] - road[i - 1];
		cnt += distance / mid;
		if (distance % mid == 0)
			cnt--;
	}
	return cnt > m;
}
int main()
{
	ios::sync_with_stdio(0);
	cin.tie(0);
	cin >> n >> m >> l;
	road.push_back(0);
	road.push_back(l);
	while (n--)
	{
		int idx;
		cin >> idx;
		road.push_back(idx);
	}
	sort(road.begin(), road.end());
	int start = 1;
	int end = l;
	while (start <= end)
	{
		int mid = (start + end) / 2;
		if (func(mid))
			start = mid + 1;
		else
		{
			ret = mid;
			end = mid - 1;
		}
	}
	cout << ret;
}

File: src/AS_24_01_week2/wangon/1477.py

def func(mid, size):
    cnt = 0
    for i in range(1, size):
      distance = road[i] - road[i - 1]
      cnt += distance // mid
      if distance % mid == 0:
        cnt -= 1
    return cnt > m

n, m, l = map(int, input().split())
road = [0, l]

values = list(map(int, input().split()))

for val in values:
   road.append(val)

road.sort()

start = 1
end = l
ret = 0

while start <= end:
    mid = (start + end) // 2
    if func(mid, len(road)) == True:
        start = mid + 1
    else:
        ret = mid
        end = mid - 1

print(ret)
